core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java

```diff
@@ -211,4 +216,15 @@ int sizeOfNormalGroupMap() {
   int sizeOfDelayedGroupMap() {
     return delayedGroupMap.size();
   }
+
+  void abortAllGroups() {
```

Comment:
I would like to understand it correctly.
Calling this is for earlier recovery and not required for safety and liveness, right?
I'm a little concerned about the `the clients will wait forever` part in the description since it sounds like no liveness is guaranteed.
There might be a case where a client crashes before aborting all groups (calling this method) so I think the correctness should be guaranteed without this.

Reply:
> Calling this is for earlier recovery and not required for safety and liveness, right?

Thanks for the question. I think this is for liveness property.

In the following special case, the client of a transaction waits forever (maybe [this example code](https://github.com/scalar-labs/scalardb/pull/1650#discussion_r1566716082) is more helpful to understand):
1. `client#1` begins `txn#1`. The NormalGroup contains `[txn#1(empty)]` now
2. `client#2` begins `txn#2`. The NormalGroup contains `[txn#1(empty), txn#2(empty)]` now
3. `txn#1` failes in the CRUD (execution) phase, and `client#1` doesn't abort it for some reason.  The NormalGroup contains `[txn#1(empty), txn#2(empty)]` now
4. `txn#2` executes some CRUD operations and try committing. The NormalGroup contains `[txn#1(empty), txn#2(snapshot-for-txn#2)]` now
5. The application closes the TransactionManager by calling `close()` which results in calling `GroupCommitter.close()`, then all the background workers stop. The NormalGroup contains `[txn#1(empty), txn#2(snapshot-for-txn#2)]` now
6. `client#2` is still waiting on the CompletableFuture inside the slot for `txn#2`. The NormalGroup contains `[txn#1(empty), txn#2(snapshot-for-txn#2)]` now
7. No background workers taking care of the NormalGroup is running. `client#2` gets stuck since the NormalGroup won't be READY due to the open slot for `txn#1`

> where a client crashes before aborting all groups (calling this method) so I think the correctness should be guaranteed without this.

In this case (the client didn't close the TransactionManager, right?), DelayedSlotMoveWorker will either move `txn#1` to a DelayedGroup or remove timed-out `txn#1` from the NormalGroup so that `txn#2` will be committed. The point is if `GroupCommitter.close()` is called.

This PR for the case a client closes the TransactionManager while there are ongoing transactions and the client keeps waiting on the transactions.

Reply:
Addressed it in https://github.com/scalar-labs/scalardb/pull/1650/commits/dd1a0bb94ddc002decce8104ddc103120d82ff5c